### PR TITLE
Add billing provider to filterable parameters for subscription resource

### PR DESCRIPTION
**Pull request by awood**
Jira issue: SWATCH-2244

## Description
Allow filtering based on billing provider in calls to `subscriptions/products/<product_id>`

### Setup
1.  Sync offerings from stage.  You can create the offerings yourself, but it's easier to just pull the stage data.  You'll need to have the keystore and truststore for the stage subscription endpoint along with the password.  You'll also need the stage subscription endpoint.  I can provide all those or you can retrieve them from the `api` pod's environment (`SUBSCRIPTION_URL`) and Vault (`/insights/show/secrets/insights-stage/rhsm-stage/pinhead` and `/insights/show/secrets/insights-stage/rhsm-stage/tls` for the password).
    ```
    http PUT :8000/api/rhsm-subscriptions/v1/internal/rpc/offerings/sync x-rh-swatch-psk:placeholder
    ```
1.  Create (or find) an org with subscriptions that have a billing provider.  I pulled in stage data for the `subscriptions` table and then used to following to randomly select an org from the results of the second query (and making my request based on the tag given by the first query)
    ```sql
    SELECT *
    FROM sku_product_tag
    WHERE sku in  (
      SELECT DISTINCT sku
      FROM subscription
      WHERE billing_provider = 'aws'
      GROUP BY sku
    );

    select org_id from subscription where sku = 'MCT4121HR' and billing_provider = 'aws';
    ```

### Steps

1.  Make a call filtering on `billing_provider`.
    ```
    http :8000/api/rhsm-subscriptions/v1/subscriptions/products/OpenShift%20Container%20Platform \
    billing_provider==azure \
    beginning==2022-03-01T00:00:00.000Z \
    ending==2024-03-31T23:59:59.999Z \
    x-rh-identity:$(echo -n '{"identity":{"account_number":"","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"ORG_HERE"}}}' | base64 -w 0)
    ```

1.  Make the same call but with no `billing_provider` filter
    ```
    http :8000/api/rhsm-subscriptions/v1/subscriptions/products/OpenShift%20Container%20Platform \
    beginning==2022-03-01T00:00:00.000Z \
    ending==2024-03-31T23:59:59.999Z \
    x-rh-identity:$(echo -n '{"identity":{"account_number":"","type":"User","user":{"is_org_admin":true},"internal":{"org_id":"ORG_HERE"}}}' | base64 -w 0)
    ```

### Verification
1.  You shouldn't see any AWS results in the first `http` call and you should see them in the second.

### CODE REVIEW

**Review by Sgitario — CHANGES_REQUESTED** [review]
Apart from the comments, the changes were incompatible with main and that's why the CI was failing. I corrected the incompatibilities myself (it was only about to add the metricId field), so CI should look better.

**Review by Sgitario — CHANGES_REQUESTED** [review]
One more comment that I didn't spot in the first round.

File: src/main/java/org/candlepin/subscriptions/resource/SubscriptionTableController.java [hunk 1]
```diff
@@ -115,7 +115,8 @@ public SkuCapacityReport capacityReportBySku( // NOSONAR
             + "Usage={}, "
             + "between={} and {}, "
             + "Uom={}, "
-            + "MetricId={}",
+            + "MetricId={}"
```
Comment (Sgitario) [inline src/main/java/org/candlepin/subscriptions/resource/SubscriptionTableController.java:118]:
```suggestion
            + "MetricId={}, "
```

**Review by Sgitario — COMMENTED** [review]
Thanks for addressing my comments. 
The changes look good to me, there is only one minor thing (commit suggestion).

**Comment by awood** [comment]
> Thanks for addressing my comments.
> The changes look good to me, there is only one minor thing (commit suggestion).

Thanks for that.  Silly error on my part.

**Review by Sgitario — APPROVED** [review]
lgtm!

**Review by ntkathole — APPROVED** [review]
Looks good!
Coverage added in https://gitlab.cee.redhat.com/insights-qe/iqe-rhsm-subscriptions-plugin/-/merge_requests/616/diffs 
```

In [6]: app.rhsm_subscriptions.get_skus_report("rhel-for-x86-els-payg", billing_provider="aws")
2024-03-19 10:50:36.336 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=None, params=[('sort', 'next_event_date'), ('limit', 100), ('offset', 0), ('billing_provider', 'aws')]
Out[6]:
[{'billing_provider': 'aws',
  'capacity': 0,
  'has_infinite_quantity': False,
  'hypervisor_capacity': 0,
  'next_event_date': datetime.datetime(2024, 3, 31, 8, 50, 29, tzinfo=tzlocal()),
  'next_event_type': 'Subscription End',
  'product_name': 'Red Hat Enterprise Linux for Third Party Linux Migration '
                  'with Extended Life Cycle Support (On-Demand, Hourly '
                  'Production Support)',
  'quantity': 1,
  'service_level': 'Premium',
  'sku': 'RH02781HR',
  'subscriptions': [{'id': '843838', 'number': '7660330'}],
  'total_capacity': 0,
  'usage': 'Production'}]

In [7]: app.rhsm_subscriptions.get_skus_report("rhel-for-x86-els-payg", billing_provider="azure")
2024-03-19 10:50:40.766 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=None, params=[('sort', 'next_event_date'), ('limit', 100), ('offset', 0), ('billing_provider', 'azure')]
Out[7]:
[{'billing_provider': 'azure',
  'capacity': 0,
  'has_infinite_quantity': False,
  'hypervisor_capacity': 0,
  'next_event_date': datetime.datetime(2024, 3, 31, 8, 50, 24, tzinfo=tzlocal()),
  'next_event_type': 'Subscription End',
  'product_name': 'Red Hat Enterprise Linux for Third Party Linux Migration '
                  'with Extended Life Cycle Support (On-Demand, Hourly '
                  'Production Support)',
  'quantity': 1,
  'service_level': 'Premium',
  'sku': 'RH02781HR',
  'subscriptions': [{'id': '802470', 'number': '7551621'}],
  'total_capacity': 0,
  'usage': 'Production'}]

In [8]: app.rhsm_subscriptions.get_skus_report("rhel-for-x86-els-payg")
2024-03-19 10:50:49.777 [    INFO] [iqe.base.rest_client] REST: METHOD=GET, request_id=None, params=[('sort', 'next_event_date'), ('limit', 100), ('offset', 0)]
Out[8]:
[{'billing_provider': '',
  'capacity': 0,
  'has_infinite_quantity': False,
  'hypervisor_capacity': 0,
  'next_event_date': datetime.datetime(2024, 3, 31, 8, 50, 24, tzinfo=tzlocal()),
  'next_event_type': 'Subscription End',
  'product_name': 'Red Hat Enterprise Linux for Third Party Linux Migration '
                  'with Extended Life Cycle Support (On-Demand, Hourly '
                  'Production Support)',
  'quantity': 2,
  'service_level': 'Premium',
  'sku': 'RH02781HR',
  'subscriptions': [{'id': '802470', 'number': '7551621'},
                    {'id': '843838', 'number': '7660330'}],
  'total_capacity': 0,
  'usage': 'Production'}]
```

**Pull request merged**